### PR TITLE
generic callback actionbars

### PR DESCRIPTION
**Pull request by Adharainspace**
## About the PR 
so a lot of times if you want to add anything that takes time, its good to use an actionbar. but you end up just adding a whole shit tonne of code for something that ends up being "click on thing. wait for 4 seconds. if the person hasn't moved away, do something". this pr adds in a generic version of that so you don't have to define new types everywhere. leads to much cleaner code. i didnt actually make anything use this though, i figured i would do that in a separate pr. has some documentation with it too, kinda. also a changes an instance of world.time to TIME.  its all been tested too. 

## Why's this needed? 
pretty much what i said above. leads to cleaner code, less weird organization things too.

### CODE REVIEW

**Comment by ZeWaka** [comment]
Can we get an addition to https://github.com/goonstation/goonstation/blob/master/.github/CODE_GUIDELINES.md#useful-things for this

File: code/datums/controllers/action_controls.dm [hunk 1]
```diff
@@ -346,6 +346,76 @@ var/datum/action_controller/actions
...
 		..()
 
+/datum/action/bar/icon/callback //calls a specified proc if it finishes without interruptions, moderately customiseable
+	id = null //set to a string version of the callback proc path
+	interrupt_flags = INTERRUPT_MOVE | INTERRUPT_ACT | INTERRUPT_STUNNED | INTERRUPT_ACTION //unsure if these will need to be overriden
+	var/proc_path = null //set to the proc path of the proc to be "called back"
+	var/target = null //what the target of the action is, if any
+	var/end_message = ""
+	var/maximum_range = 1 //what is the maximum range target and owner can be apart? need to modify directly before starting the action.
+	var/list/proc_args = null //a list of args for the callback proc, if needed.
```
Comment (ZeWaka) [inline code/datums/controllers/action_controls.dm:356]: can you switch these to use the doc comment format

File: _std/macros/actions.dm [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+/* example for alternate use:
+
+/obj/item/foo
+	name = "foo"
+	icon = 'icons/obj/foo.dmi'
+	icon_state = "foo"
+
+	attack_self(var/mob/M)
+		M.visible_message("[M] starts fiddling with \the [src].")
+		var/datum/action/bar/icon/callback/action_bar = new /datum/action/bar/icon/callback(M, src, 5 SECONDS, null, /obj/item/foo/proc/cool_proc,\
```
Comment (ZeWaka) [inline _std/macros/actions.dm:37]: this example doesn't seem to actually use the macro?

File: code/datums/controllers/action_controls.dm [hunk 3]
```diff
@@ -346,6 +346,76 @@ var/datum/action_controller/actions
...
+
+	New(var/owner, var/target, var/duration, var/owner_proc_path, var/target_proc_path, var/icon, var/icon_state, var/end_message)
```
Comment (ZeWaka) [inline code/datums/controllers/action_controls.dm:358]: recording re: imcoder discussion, only one proc path is needed

**Comment by Adharainspace** [comment]
i forgot tags ffs

File: code/datums/controllers/action_controls.dm (hunk 1 above)
Comment (Adharainspace) [reply code/datums/controllers/action_controls.dm]: done!

**Comment by Adharainspace** [comment]
> Can we get an addition to https://github.com/goonstation/goonstation/blob/master/.github/CODE_GUIDELINES.md#useful-things for this

done!

File: code/datums/controllers/action_controls.dm (hunk 3 above)
Comment (Adharainspace) [reply code/datums/controllers/action_controls.dm]: done!

File: _std/macros/actions.dm (hunk 2 above)
Comment (Adharainspace) [reply _std/macros/actions.dm]: clarified why i put it there!

File: code/datums/controllers/action_controls.dm [hunk 4]
```diff
@@ -104,7 +104,7 @@ var/datum/action_controller/actions
 
 	proc/onRestart()			   //Called when the action restarts (for example: automenders)
 		sleep(1)
-		started = world.time
+		started = TIME
```
Comment (pali6) [inline code/datums/controllers/action_controls.dm:107]: Won't this need changes in other places too?

Comment (Adharainspace) [reply code/datums/controllers/action_controls.dm:107]: it did, done!

File: .github/CODE_GUIDELINES.md [hunk 5]
```diff
@@ -205,3 +205,6 @@ If total cpu and real time are the same the proc never sleeps, otherwise real ti
 
 ### Signals and Components
 * ninjanomnom from TG has written up a [useful primer](https://tgstation13.org/phpBB/viewtopic.php?f=5&t=22674) on signals and components. Most of the stuff there applies, although elements do not exist in this codebase.
+
+### Generic Action bar
+* Hate coding action bars? Making a new definition for an action bar datum just so you have visual feedback for your construction feel gross? Well fear not! You can now use the SETUP_GENERIC_ACTIONBAR() macro! Check [_std/macros/actions.dm`](https://github.com/goonstation/goonstation/blob/681d0b0a28ede4f0b7a09f02cf9ff79bb0b23b7c/_std/macros/actions.dm) for more information.
```
Comment (ZeWaka) [inline .github/CODE_GUIDELINES.md:210]:
```suggestion
* Hate coding action bars? Making a new definition for an action bar datum just so you have visual feedback for your construction feel gross? Well fear not! You can now use the SETUP_GENERIC_ACTIONBAR() macro! Check [_std/macros/actions.dm`](https://github.com/goonstation/goonstation/blob/master/_std/macros/actions.dm) for more information.  
```

File: _std/macros/actions.dm [hunk 6]
```diff
@@ -0,0 +1,45 @@
+/**
+* uses /datum/action/bar/icon/callback, in code/datums/controllers/action_controls.dm
```
Comment (ZeWaka) [inline _std/macros/actions.dm:2]:
```suggestion
* uses [/datum/action/bar/icon/callback]
```

File: code/datums/controllers/action_controls.dm [hunk 7]
```diff
@@ -346,6 +346,85 @@ var/datum/action_controller/actions
 			del(icon_image)
 		..()
 
+
+/**
+* calls a specified proc if it finishes without interruptions.
+* check [_std/macros/actions.dm](https://github.com/goonstation/goonstation/blob/681d0b0a28ede4f0b7a09f02cf9ff79bb0b23b7c/_std/macros/actions.dm) for macro documentation
```
Comment (ZeWaka) [inline code/datums/controllers/action_controls.dm:352]:
```suggestion
* check [_std/macros/actions.dm] for macro documentation
```

File: code/datums/controllers/action_controls.dm [hunk 8]
```diff
@@ -346,6 +346,85 @@ var/datum/action_controller/actions
 			del(icon_image)
 		..()
 
+
+/**
+* calls a specified proc if it finishes without interruptions.
```
Comment (ZeWaka) [inline code/datums/controllers/action_controls.dm:351]:
```suggestion
* calls a specified proc if it finishes without interruptions.
*
```

Comment (Adharainspace) [reply code/datums/controllers/action_controls.dm:351]: done!

File: code/datums/controllers/action_controls.dm (hunk 7 above)
Comment (Adharainspace) [reply code/datums/controllers/action_controls.dm]: done!

File: _std/macros/actions.dm (hunk 6 above)
Comment (Adharainspace) [reply _std/macros/actions.dm]: done!

File: .github/CODE_GUIDELINES.md (hunk 5 above)
Comment (Adharainspace) [reply .github/CODE_GUIDELINES.md]: done!

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**